universal-application-tool-0.0.1/app/filters/LoggingFilter.java

```diff
@@ -0,0 +1,43 @@
+package filters;
+
+import java.time.Clock;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import play.mvc.EssentialAction;
+import play.mvc.EssentialFilter;
+
+/** This is a simple filter that produces an apache-style access log. */
+@Singleton
+public class LoggingFilter extends EssentialFilter {
+
+  private final Executor exec;
+  private final Clock clock;
+  private static final Logger log = LoggerFactory.getLogger("loggingfilter");
+
+  /** @param exec This class is needed to execute code asynchronously. */
+  @Inject
+  public LoggingFilter(Executor exec, Clock clock) {
+    this.exec = exec;
+    this.clock = clock;
+  }
+
+  @Override
+  public EssentialAction apply(EssentialAction next) {
+    return EssentialAction.of(
+        request -> {
+          final long startTime = clock.millis();
+          return next.apply(request)
+              .map(
+                  result -> {
+                    long time = clock.millis() - startTime;
+                    log.info(
+                        "{}\t{}\t{}ms\t{}", request.method(), request.uri(), time, result.status());
+                    return result.withHeader("X-RequestTime", String.valueOf(time));
```

Comment:
**Question**: Is there a specific string format or a predefined list of strings this (`X-RequestTime`) should follow? Or can this be an arbitrary string, i.e. it is not defined or used anywhere else?

Reply:
Yep, HTTP headers can be anything.  Traditionally nonstandard headers are prefixed with `X-`.

Reply:
Whoops - I looked it up and actually this convention was recommended until the mid 2010s and now it's not recommended.  We still use it at Google sometimes.  I'll remove it!